HeterogeneousCore/CUDAUtilities/interface/HistoContainerAlgo.h

```diff
@@ -0,0 +1,172 @@
+      grid.sync();
+      finalizeCoop(view, ws);
+      grid.sync();
+      countOrFillFromVector<Histo, T, CountOrFill::fill>(h, nh, v, offsets);
+    }
+#endif
+
+    template <typename Histo, typename T>
+    inline __attribute__((always_inline)) void fillManyFromVectorCoop(Histo *h,
+                                                                      uint32_t nh,
+                                                                      T const *v,
+                                                                      uint32_t const *offsets,
+                                                                      int32_t totSize,
+                                                                      int nthreads,
+                                                                      typename Histo::index_type *mem,
+                                                                      cudaStream_t stream
+#ifndef __CUDACC__
+                                                                      = cudaStreamDefault
+#endif
+    ) {
+      using View = typename Histo::View;
+      View view = {h, nullptr, mem, -1, totSize};
+#ifdef __CUDACC__
+      auto kernel = fillManyFromVectorCoopKernel<Histo, T>;
+      auto nblocks = (totSize + nthreads - 1) / nthreads;
+      assert(nblocks > 0);
+      auto nOnes = view.size();
+      auto nchunks = nOnes / nthreads + 1;
+      auto ws = cms::cuda::make_device_unique<typename View::Counter[]>(nchunks, stream);
+      auto wsp = ws.get();
+      // FIXME: discuss with FW team: cuda calls are expensive and not needed for each event
+      static int maxBlocks = maxCoopBlocks(kernel, nthreads, 0, 0);
+      auto ncoopblocks = std::min(nblocks, maxBlocks);
+      assert(ncoopblocks > 0);
+      void *kernelArgs[] = {&view, &nh, &v, &offsets, &totSize, &wsp};
+      dim3 dimBlock(nthreads, 1, 1);
+      dim3 dimGrid(ncoopblocks, 1, 1);
+      // launch
+      cudaCheck(cudaLaunchCooperativeKernel((void *)kernel, dimGrid, dimBlock, kernelArgs, 0, stream));
+      cudaCheck(cudaGetLastError());
+#else
+      launchZero(view, stream);
+      countFromVector(h, nh, v, offsets);
+      h->finalize();
+      fillFromVector(h, nh, v, offsets);
+#endif
+    }
+
+    // iteratate over N bins left and right of the one containing "v"
+    template <typename Hist, typename V, typename Func>
+    __host__ __device__ __forceinline__ void forEachInBins(Hist const &hist, V value, int n, Func func) {
+      int bs = Hist::bin(value);
+      int be = std::min(int(Hist::nbins() - 1), bs + n);
+      bs = std::max(0, bs - n);
+      assert(be >= bs);
+      for (auto pj = hist.begin(bs); pj < hist.end(be); ++pj) {
+        func(*pj);
+      }
+    }
+
+    // iteratate over bins containing all values in window wmin, wmax
+    template <typename Hist, typename V, typename Func>
+    __host__ __device__ __forceinline__ void forEachInWindow(Hist const &hist, V wmin, V wmax, Func const &func) {
```

Comment:
The `forEachInBins()` above took similar `func` argument by value. I wonder which would make most sense. Perhaps
```suggestion
    __host__ __device__ __forceinline__ void forEachInWindow(Hist const &hist, V wmin, V wmax, Func &&func) {
```
? Or is `func` required be thread safe?